### PR TITLE
SIMCoupledSI: add support for relaxation with optional aitken acceleration

### PR DESCRIPTION
**Pull request by akva2**

### CODE REVIEW

File: Apps/Common/SIMCoupledSI.h [hunk 1]
```diff
@@ -62,6 +85,29 @@ class SIMCoupledSI : public SIMCoupled<T1,T2>
 
       if ((conv = this->checkConvergence(tp,status1,status2)) <= SIM::DIVERGED)
         return false;
+
+      // Calculate aitken acceleration factor
+      if (aitken && omega0 != 0.0) {
```
Comment (kmokstad) [inline Apps/Common/SIMCoupledSI.h:90]: The bool `aitken` is strictly not needed. Can just use `omega0 != 0.0`to flag whether the Aitken relaxation is to be used or not.

File: Apps/Common/SIMCoupledSI.h [hunk 2]
```diff
@@ -62,6 +85,29 @@ class SIMCoupledSI : public SIMCoupled<T1,T2>
...
+        if (tp.iter > 0) {
+            Vector r1 = this->getAitkenResidual();
+            r1 -= prevRes;
+            omega *= -prevRes.dot(r1) / r1.dot(r1);
+        }
+        prevRes = this->getAitkenResidual();
+      }
+
+      // Perform relaxation
+      if (omega0 != 0.0) {
```
Comment (kmokstad) [inline Apps/Common/SIMCoupledSI.h:100]: This if is not necessary - the block can be included in the previous `if (omega0 != 0.0)`

File: Apps/Common/SIMCoupledSI.h [hunk 3]
```diff
@@ -90,6 +136,11 @@ class SIMCoupledSI : public SIMCoupled<T1,T2>
 
 protected:
   int maxIter; //!< Maximum number of iterations
+  double omega; //!< Relaxation parameter
+  double omega0; //!< Initial relaxation parameter
+  bool aitken; //!< True to enable aitken-acceleration
```
Comment (kmokstad) [inline Apps/Common/SIMCoupledSI.h:141]: Necessary? (see above)

File: Apps/Common/SIMCoupledSI.h [hunk 4]
```diff
@@ -62,6 +85,29 @@ class SIMCoupledSI : public SIMCoupled<T1,T2>
 
       if ((conv = this->checkConvergence(tp,status1,status2)) <= SIM::DIVERGED)
         return false;
+
+      // Calculate aitken acceleration factor
+      if (aitken && omega0 != 0.0) {
+        if (tp.iter > 0) {
+            Vector r1 = this->getAitkenResidual();
+            r1 -= prevRes;
```
Comment (kmokstad) [inline Apps/Common/SIMCoupledSI.h:93]: `Vector r1 = this->getAitkenResidual() - prevRes;` should also work.

**Review by kmokstad — COMMENTED** [review]
Looks OK, except small things. Haven't checked downstreams yet, so I might have missed some points.

File: Apps/Common/SIMCoupledSI.h (hunk 1 above)
Comment (akva2) [reply Apps/Common/SIMCoupledSI.h:90]: it is, you can do relaxation without aitken.

File: Apps/Common/SIMCoupledSI.h (hunk 2 above)
Comment (akva2) [reply Apps/Common/SIMCoupledSI.h:100]: no, because you can do relaxation without aitken.

File: Apps/Common/SIMCoupledSI.h (hunk 3 above)
Comment (akva2) [reply Apps/Common/SIMCoupledSI.h:141]: yes, see above.

File: Apps/Common/SIMCoupledSI.h (hunk 4 above)
Comment (akva2) [reply Apps/Common/SIMCoupledSI.h:93]: which yields an additional copy. granted, very minor optimization but still i prefer the current code for that reason.

Comment (kmokstad) [reply Apps/Common/SIMCoupledSI.h:93]: ok, keep yours

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-advectiondiffusion=54 ifem-stokes=165 ifem-navierstokes=135 ifem-openfrac=59 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-advectiondiffusion=54 ifem-stokes=165 ifem-navierstokes=135 ifem-openfrac=59 ifem-boussinesq=48 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-advectiondiffusion=54 ifem-stokes=165 ifem-navierstokes=135 ifem-openfrac=59 ifem-boussinesq=48 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-advectiondiffusion=54 ifem-stokes=165 ifem-navierstokes=135 ifem-openfrac=59 ifem-boussinesq=48 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-advectiondiffusion=54 ifem-stokes=165 ifem-navierstokes=135 ifem-openfrac=59 ifem-boussinesq=48 please

**Pull request merged**